### PR TITLE
[llvm-objcopy] Add --remove-symbol-prefix

### PR DESCRIPTION
**Pull request by kongy**
llvm-objcopy has functionality to add prefixes to symbols (--prefix-symbols). This adds an inverse action to undo the modification and return to the original.

### CODE REVIEW

File: llvm/tools/llvm-objcopy/ObjcopyOpts.td
Comment (jh7370) [inline llvm/tools/llvm-objcopy/ObjcopyOpts.td:207]: I feel like `remove-symbol-prefix` would be an easier-to-read name (but if you're aiming for compatibility with something else, please say so).

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp
Comment (jh7370) [inline llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:741]: Why this error? It seems to me like you could do the prefix removal first and then the addition of the prefix, e.g. with removing "foo" and adding "bar", you could change "fooblob" into "barblob" with one llvm-objcopy run.

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:3]: llvm-readelf will provide a more compact output, making it easier to see what's actually interesting about this.

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:33]: This isn't going to do what you think it's going to do. Please re-read the FileCheck documentation and take a look at other tests (hint: NEXT is a suffix, not a pattern).

File: llvm/include/llvm/ObjCopy/CommonConfig.h
Comment (jh7370) [inline llvm/include/llvm/ObjCopy/CommonConfig.h:221]: Please make sure that non-ELF formats complain when this option is used, in the same manner as `SymbolsPrefix`. See `ConfigManager.cpp` for this.

**Review by jh7370 — CHANGES_REQUESTED** [review]
Hi, thanks for the patch! What's the motivation behind being able to remove prefixes from symbol names? Note that removing prefixes could lead to various issues, e.g. multiple symbols with the same name in the same file, resulting in the linker not knowing the right symbol to pick, so adding this option needs to be done carefully. Does GNU objcopy have any prior art for this option?

In addition to your code changes, please update the llvm-objcopy CommandGuide doc (llvm/docs/CommandGuide/llvm-objcopy.rst) to describe the new option.

**Comment by MaskRay** [comment]
GNU objcopy does not support the option. I filed https://sourceware.org/bugzilla/show_bug.cgi?id=16773

If an object file has `a` and `za`, `llvm-objcopy --prefix-symbols-remove=z` would leave to two `a`. If either one is STB_LOCAL, it should be fine. If both are STB_GLOBAL, the linker will likely report a duplicate definition error.

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (MaskRay) [reply llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:33]: The tabular output of `llvm-readelf -s` is likely more readable

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp
Comment (kongy) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:741]: Because it can be error prone. We do not specify the ordering of adding prefixes and removing prefix. The results can be different with a different order. If the user intentionally needs both actions, they can call llvm-objcopy twice.

File: llvm/tools/llvm-objcopy/ObjcopyOpts.td
Comment (kongy) [reply llvm/tools/llvm-objcopy/ObjcopyOpts.td:207]: > What's the motivation behind being able to remove prefixes from symbol names?

One of our projects prefixes its symbols with llvm-objcopy using `--prefix-symbols`. When we deploy sampling pgo, we need to undo that and get the original symbol names.

> Note that removing prefixes could lead to various issues, e.g. multiple symbols with the same name in the same file, resulting in the linker not knowing the right symbol to pick, so adding this option needs to be done carefully.

Since our goal is to undo the `--prefix-symbols` action, it is guarenteed that the output symbol names will not duplicate.

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp
Comment (appujee) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:741]: Yeah it is better to not facilitate both options at the same time. Picking the last flag could be another option but erroring out is also fine.

Comment (jh7370) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:741]: Sorry, I disagree. There is a standard model for handling multiple options in llvm-objcopy (e.g. `--remove-section` and then `--add-section`) which is to perform the subtractive option first and then do the additive one, regardless of the order in which they appear on the command-line, and this isn't significantly different to that style. In other words, in this case you'd apply the (either last value or all, I'd actually consider doing all, in order) prefix remove option before doing the prefix add option. You can mention this in the documentation and enforce the ordering by tests. Honestly, I don't think there's any risk of confusion either, since it's fairly obvious what has happened by simply looking at the names.

**Review by jh7370 — COMMENTED** [review]
> If an object file has `a` and `za`, `llvm-objcopy --prefix-symbols-remove=z` would leave two `a`. If either one is STB_LOCAL, it should be fine. If both are STB_GLOBAL, the linker will likely report a duplicate definition error.

Thinking more about this, I think it's fair to allow it (it's not the first option where people can break their object files after all).

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (kongy) [reply llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:3]: Will rewrite this (as well as `prefix-symbols.test`) in a follow-up change.

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp
Comment (kongy) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp]: Done. Added test to enforce the ordering.

File: llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp
Comment (jh7370) [inline llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp:332]: @MaskRay, I think this ordering looks about as good as any other to me, with one possible exception, but wanted a second opinion: is this the right place in the ordering for the symbol prefix removing functionality?

I think it makes sense to do options that modify a named symbol before renaming it (this is consistent with the general symbol renaming option immediately above). I think there's an unfortunate potential for confusion when also renaming symbols, as the prefix removal happens after the renaming, but at the same time, I'm not sure if it would be any less confusing if we did it the other way around.

Case 1 (current behaviour with this patch, --rename-symbols (option name and syntax possibly paraphrased - I'm too lazy to look it up) first, then --remove-prefix-symbols). In this case if you did `--rename-symbols foobar=bazbar` and `--remove-prefix-symbols=baz`, you'd end up with a symbol called `bar`. If you actually wanted a final symbol called `baz` you'd need to do the rename to `bazbazbar`.

Case 2 (do the options in the other order): if you did `--rename-symbols foobar=bazbar` and `--remove-prefix-symbols=foo`, you'd end up with a symbol named `bar`, since the prefix would have been removed before the renaming.

In other words, both orderings have some issues.

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (jh7370) [reply llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:3]: Please don't defer improvements to new tests until a separate change. Please fix this test in this PR. You can address the existing test in a separate PR: the two are not tied together.

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:14]: - [ ] I think I'd prefer to see a functional rename using this, i.e. where you remove one prefix and add a different one. For example, you could remove `foo` from `foobar` and add `baz` and then check the llvm-readelf output to see what the final result looks like (it should be `bazbar`).

File: llvm/tools/llvm-objcopy/ObjcopyOpts.td
Comment (jh7370) [reply llvm/tools/llvm-objcopy/ObjcopyOpts.td]: I don't think my suggestion for the option name has been addressed? It looks like you've just renamed the option to another name.

**Review by jh7370 — COMMENTED** [review]
@kongy, per LLVM review guidance, please don't do rebase and force push on an in-progress review, because this prevents diffing against the version that was originally reviewed, making it harder to see what's changed between the original and current. Prefer fixup commits, or if a rebase is really necessary, do a merge from main instead. Also, please don't click "Resolve Conversation" on threads I'm involved in, as I use this functionality to confirm that I have reviewed any requested changes and am happy with them. I will assume, unless stated otherwise, that you have attempted to resolve all requested changes without you needing to mark the conversation as resolved whenver you update a PR (and if you haven't, per LLVM guidance, you need to say so).

**Comment by kongy** [comment]
> @kongy, per LLVM review guidance, please don't do rebase and force push on an in-progress review, because this prevents diffing against the version that was originally reviewed, making it harder to see what's changed between the original and current. Prefer fixup commits, or if a rebase is really necessary, do a merge from main instead. Also, please don't click "Resolve Conversation" on threads I'm involved in, as I use this functionality to confirm that I have reviewed any requested changes and am happy with them. I will assume, unless stated otherwise, that you have attempted to resolve all requested changes without you needing to mark the conversation as resolved whenver you update a PR (and if you haven't, per LLVM guidance, you need to say so).

Got it. Sorry this is my first time uploading an LLVM change after the Github switch.

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (kongy) [reply llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test]: Done.

File: llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp
Comment (MaskRay) [reply llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp:332]: We have similar confusion wrt `{add,rename,remove}-section` and I don't think there is a good way to completely remove confusion...

GNU objcopy does not make these operations form a total order, which is probably worse (i.e. `-A -B`=>op A runs before B. `-B -C` => B runs before C;  `-C -A` => C runs before A; where A/B/C refer to rename-section/add-section/set-section-flags.)

I am happy with the the current order. Users are expected to run llvm-objcopy multiple times if they are confused:)

**Comment by MaskRay** [comment]
Can you update `llvm/docs/CommandGuide/llvm-objcopy.rst`?

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (MaskRay) [inline llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:11]: Did you forget `--prefix-symbols` as a test?

**Comment by MaskRay** [comment]
> title: Add functionality to llvm-objcopy to remove prefixes

Perhaps `[llvm-objcopy] Add --remove-symbol-prefix` ? (ensure that the option name is included in the commit message)

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (kongy) [reply llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:11]: Done

**Review by MaskRay — APPROVED** [review]
LGTM when jh7370 is happy

**Comment by kongy** [comment]
I'm going to merge this as no objection is raised for three days. Happy to address any further comments in follow-up CLs.

File: llvm/docs/CommandGuide/llvm-objcopy.rst
Comment (jh7370) [inline llvm/docs/CommandGuide/llvm-objcopy.rst:123]:
```suggestion
Remove ``<prefix>`` from the start of every symbol name. No-op for symbols that do
not start with ``<prefix>``.
```
Use the same formatting as the other options in this doc.

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:6]: Comment needs updating for new option name.

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:14]: I've already said this once: I don't think this is the best way of testing this, in part because I don't think it's particularly clear. I think instead you should use different strings for the prefix removal and addition and use llvm-readelf to show that the expected symbol names are in the output.

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:64]: I think you can just get away with checking the symbol names here, and don't need to check the other columns (which as an aside, feel somewhat arbitrary in how you've checked them, e.g. why `[[#]]` for Value, but explicitly 0 for Size?).
```suggestion
# CHECK-NEXT: Name
# CHECK-NEXT: {{ $}}
# CHECK-NEXT: __pf_foo
# CHECK-NEXT: bar
# CHECK-NEXT: foobar
# CHECK-NEXT: foo__pf_bar1
# CHECK-NEXT: foo__pf_bar2
# CHECK-NEXT: undef
```

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test:56]: I don't see any reason to not match the exact number of symbols here, so replace this `[[#]]` with 7.

**Comment by jh7370** [comment]
> GNU objcopy does not support the option. I filed https://sourceware.org/bugzilla/show_bug.cgi?id=16773

That bug reference seems incorrect @MaskRay?

**Review by jh7370 — COMMENTED** [review]
> I'm going to merge this as no objection is raised for three days. Happy to address any further comments in follow-up CLs.

In general, I don't think this is within LLVM policy. @MaskRay explicitly said his approval was conditional on my approval, so landing it without my approval or a revised approval from @MaskRay would effectively mean landing this without any approval at all. I've been a) off sick and b) have a high load of reviews that need to be tackled, so responding to updates takes time (even if I hadn't been off sick).

File: llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test
Comment (kongy) [reply llvm/test/tools/llvm-objcopy/ELF/prefix-symbols-remove.test]: Done.

**Review by jh7370 — APPROVED** [review]
LGTM, thanks. Do you have commit access?

**Comment by kongy** [comment]
> LGTM, thanks. Do you have commit access?

Thanks!

**Pull request merged**
